nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
+                .filter(dataType -> !dataType.getFieldType().equals(RecordFieldType.STRING));
+
+        if (value instanceof String) {
+            mostSuitableType = findMostSuitableTypeByStringValue((String) value, types, dataTypeMapper);
+        } else if (inferredDataTypeOptional.isPresent()) {
+            DataType inferredDataType = inferredDataTypeOptional.get();
+
+            Optional<T> inferredTypeOptional = types.stream()
+                    .filter(type -> dataTypeMapper.apply(type).equals(inferredDataType))
+                    .findFirst();
+
+            if (inferredTypeOptional.isPresent()) {
+                mostSuitableType = inferredTypeOptional;
+            } else {
+                Optional<T> widerAvailableTypeOptional = types.stream()
+                        .map(type -> getWiderType(dataTypeMapper.apply(type), inferredDataType).isPresent() ? type : null)
+                        .filter(Objects::nonNull)
+                        .findFirst();
+
+                if (widerAvailableTypeOptional.isPresent()) {
+                    mostSuitableType = widerAvailableTypeOptional;
+                } else {
+                    mostSuitableType = Optional.empty();
+                }
+            }
+        } else {
+            mostSuitableType = Optional.empty();
+        }
+
+        return mostSuitableType;
+    }
+
+    public static <T> Optional<T> findMostSuitableTypeByStringValue(String valueAsString, List<T> types, Function<T, DataType> dataTypeMapper) {
+        Optional<T> mostSuitableType = types.stream()
+                // Sorting based on the RecordFieldType enum ordering looks appropriate here as we want simpler types
+                //  first and the enum's ordering seems to reflect that
+                .sorted((type1, type2) -> {
+                            int comparison;
+
+                            RecordFieldType dataType1 = dataTypeMapper.apply(type1).getFieldType();
+                            RecordFieldType dataType2 = dataTypeMapper.apply(type2).getFieldType();
+
+                            // Moving TIMESTAMP at the front (at least it should precede DATE)
+                            if (dataType1 == RecordFieldType.TIMESTAMP) {
+                                comparison = -1;
+                            } else if (dataType2 == RecordFieldType.TIMESTAMP) {
+                                comparison = 1;
+                            } else {
+                                comparison = dataType1.compareTo(dataType2);
+                            }
+
+                            return comparison;
+                        }
+                )
+                .filter(type -> {
+                    boolean compatible;
+
+                    try {
+                        compatible = isCompatibleDataType(valueAsString, dataTypeMapper.apply(type));
+                    } catch (Exception e) {
+                        compatible = false;
```

Comment:
I'm not sure why any Exception would be thrown from this method. If it is, we should at least be logging it, rather than simply swallowing the Exception.

Reply:
Indeed no exception should be thrown and probably won't.

I added the try-catch for the general idea that during iterating through all the types, if for whatever reason a non-fitting type behaves strangely and throws an exception it shouldn't abruptly stop the selection process (as the correct type is very likely coming later after the bad-behaving one).

I thought that knowing why a given non-fitting type would throw an exception is not too important.
For example it might be possible that trying to see if a regular string or a number is a timestamp or a date, would involve something  - maybe later on, if not now - that would cause some kind of formatting related exception. We probably wouldn't care about it that much as long as we find the proper type.

But I can add a log line here. It shouldn't happen and if it does it might be useful to know.

Reply:
Added error log.